packages/eslint-plugin/src/rules/no-wrapper-object-types.ts

```diff
@@ -0,0 +1,82 @@
+import type { TSESLint, TSESTree } from '@typescript-eslint/utils';
+import { AST_NODE_TYPES } from '@typescript-eslint/utils';
+
+import { createRule } from '../util';
+
+const classNames = new Set([
+  'BigInt',
+  // eslint-disable-next-line @typescript-eslint/internal/prefer-ast-types-enum
+  'Boolean',
+  'Function',
+  'Number',
+  'Object',
+  // eslint-disable-next-line @typescript-eslint/internal/prefer-ast-types-enum
+  'String',
+  'Symbol',
+]);
+
+export default createRule({
+  name: 'no-wrapper-object-types',
+  meta: {
+    type: 'problem',
+    docs: {
+      description: 'Disallow using confusing built-in primitive class wrappers',
+      recommended: 'recommended',
+    },
+    fixable: 'code',
+    messages: {
+      bannedClassType:
+        'Prefer using the primitive `{{preferred}}` as a type name, rather than the class wrapper `{{typeName}}`.',
+      bannedFunctionType: [
+        'The `Function` type accepts any function-like value.',
+        'It provides no type safety when calling the function, which can be a common source of bugs.',
+        'It also accepts things like class declarations, which will throw at runtime as they will not be called with `new`.',
+        'If you are expecting the function to accept certain arguments, you should explicitly define the function shape.',
```

Comment:
I wonder - is there a good doc on options here? Eg defining `(...args: any) => any` or something?

Reply:
I'm worried about users seeing it and thinking it's tailored to the specific function type.

Reply:
https://github.com/typescript-eslint/typescript-eslint/pull/9102#discussion_r1623618173